plugins/vendor/zoom/ZoomDropFolderPlugin/batch/KZoomDropFolderEngine.php

```diff
@@ -0,0 +1,423 @@
+							{
+								$isParentEntry = true;
+								$this->addDropFolderFile($meetingFile, $recordingFile, $entry->id, true);
+							}
+							else
+							{
+								$this->addDropFolderFile($meetingFile, $recordingFile, $entry->id);
+							}
+							self::$lastHandledMeetingTime = self::convertTimeToUnix($meetingFile[self::START_TIME]);
+						}
+					}
+					else
+					{
+						$dropFolderFile = $dropFolderFilesMap[$recordingFileName];
+						unset($dropFolderFilesMap[$recordingFileName]);
+						$this->handleExistingDropFolderFile($dropFolderFile);
+					}
+				}
+			}
+		}
+	}
+	
+	protected function getEntryByReferenceId($referenceId)
+	{
+		try
+		{
+			$entryFilter = new KalturaBaseEntryFilter();
+			$entryFilter->referenceIdEqual = $referenceId;
+			
+			$entryPager = new KalturaFilterPager();
+			$entryPager->pageSize = 1;
+			$entryPager->pageIndex = 1;
+			KBatchBase::impersonate($this->dropFolder->partnerId);
+			$entryList = KBatchBase::$kClient->baseEntry->listAction($entryFilter, $entryPager);
+			KBatchBase::unimpersonate();
+			if (is_array($entryList->objects) && isset($entryList->objects[0]) )
+			{
+				return $entryList->objects[0];
+			}
+			
+			return null;
+		}
+		catch (Exception $e)
+		{
+			KalturaLog::err('Failed to get entry by reference id: '. $referenceId . $e->getMessage() );
+			return null;
+		}
+	}
+	
+	protected function createEntry($uuid)
+	{
+		$newEntry = new KalturaMediaEntry();
+		$newEntry->sourceType = KalturaSourceType::URL;
+		$newEntry->mediaType = KalturaMediaType::VIDEO;
+		$newEntry->referenceId = zoomProcessor::ZOOM_PREFIX . $uuid;
+		KBatchBase::impersonate($this->dropFolder->partnerId);
+		$entry = KBatchBase::$kClient->baseEntry->add($newEntry);
+		KBatchBase::unimpersonate();
+		return $entry;
+	}
+	
+	protected static function orderRecordingFiles($recordingFiles)
+	{
+		$recordingFilesOrdered = array();
+		foreach($recordingFiles as $recordingFile)
+		{
+			if(!isset($recordingFilesOrdered[$recordingFile[self::RECORDING_START]]))
+			{
+				$recordingFilesOrdered[$recordingFile[self::RECORDING_START]] = array();
+			}
+			$recordingFilesOrdered[$recordingFile[self::RECORDING_START]][] = $recordingFile;
+		}
+		ksort($recordingFilesOrdered);
+		return self::orderRecordingFilesByType($recordingFilesOrdered);
+	}
+	
+	protected static function orderRecordingFilesByType($recordingFilesOrdered)
+	{
+		foreach ($recordingFilesOrdered as $time => $recordingFilesPerTimeSlot)
+		{
+			$filesOrderByType = array();
+			foreach ($recordingFilesPerTimeSlot as $recordingFile)
+			{
+				if(!isset($filesOrderByType[$recordingFile[self::RECORDING_FILE_TYPE]]))
+				{
+					$filesOrderByType[$recordingFile[self::RECORDING_FILE_TYPE]] = array();
+				}
+				$filesOrderByType[$recordingFile[self::RECORDING_FILE_TYPE]][] = $recordingFile;
+			}
+			ksort($filesOrderByType);
+			$byTime = array();
+			foreach ($filesOrderByType as $fileOrderByType)
+			{
+				foreach ($fileOrderByType as $file)

```

Comment:
this is 3 level of for loop, what can be the worst complexity here?

Reply:
worst complexity: num of time slot * 4 * 4

first loop is by time slot - the num of times the user did start->stop->start in a meeting- I guess its only a few
second loop is by type - there are only 4 types : audio, video, chat, transcription.
third  loop is by recordings of every type - for example going over all the chat recordings - I think it wont be more than 4 for a type.